### PR TITLE
🌱 Fix self-hosted flakes in E2E tests

### PR DESCRIPTION
**Pull request by fabriziopandini**
**What this PR does / why we need it**:
This PR aims to fix self-hosted flakes by ensuring that mhc is not run during this test and by adding two delay in order to avoid doing move too aggressively.

In order to make this happen the entire management of cluster templates was re-architected so:
- cluster templates for tests are now generated from a set composable bases
- MHC-remediation has a new dedicated template
- all the other templates are without MHC
- also, in order to simplify things, this PR drops the distinction between CI and DEV test config

On top of that, this PR includes, a set of nits for improving test logs (mostly uppercase)

**Which issue(s) this PR fixes**:
Fixes https://github.com/kubernetes-sigs/cluster-api/issues/3589

### CODE REVIEW

File: test/e2e/data/infrastructure-docker/bases/cluster-with-kcp.yaml [hunk 1]
```diff
@@ -0,0 +1,70 @@
+---
```
Comment (fabriziopandini) [inline test/e2e/data/infrastructure-docker/bases/cluster-with-kcp.yaml:1]: I would like to gitignore the generated cluster-templates and keep under source control only the bases, but it seems that this makes the apidiff batch to fail 🤔

File: test/e2e/data/infrastructure-docker/cluster-template-kcp-adoption.yaml [hunk 2]
```diff
@@ -1,179 +1,180 @@
...
-# these resources are sequenced by label:
-# 1. initial: just the resources to bootstrap an initial controlplane
-# 2. kcp: the KCP resources (note the duplicated Cluster to associate the control plane)
-##
-###
-# 1. initial
-###
-apiVersion: infrastructure.cluster.x-k8s.io/v1alpha3
-kind: DockerCluster
+apiVersion: v1
```
Comment (fabriziopandini) [inline test/e2e/data/infrastructure-docker/cluster-template-kcp-adoption.yaml:1]: I would like to gitignore the generated cluster-templates and keep under source control only the bases, but it seems that this makes the apidiff batch to fail 🤔

File: test/e2e/data/infrastructure-docker/cluster-template-mhc.yaml [hunk 3]
```diff
@@ -1,147 +1,154 @@
...
-spec:
-  template:
-    spec:
-      extraMounts:
-        - containerPath: "/var/run/docker.sock"
-          hostPath: "/var/run/docker.sock"
----
-kind: KubeadmControlPlane
-apiVersion: controlplane.cluster.x-k8s.io/v1alpha3
+apiVersion: v1
```
Comment (fabriziopandini) [inline test/e2e/data/infrastructure-docker/cluster-template-mhc.yaml:1]: I would like to gitignore the generated cluster-templates and keep under source control only the bases, but it seems that this makes the apidiff batch to fail 🤔

File: test/e2e/data/infrastructure-docker/cluster-template.yaml [hunk 4]
```diff
@@ -1,147 +1,139 @@
-apiVersion: infrastructure.cluster.x-k8s.io/v1alpha3
-kind: DockerCluster
+apiVersion: v1
```
Comment (fabriziopandini) [inline test/e2e/data/infrastructure-docker/cluster-template.yaml:1]: I would like to gitignore the generated cluster-templates and keep under source control only the bases, but it seems that this makes the apidiff batch to fail 🤔

File: test/e2e/config/docker-dev.yaml [hunk 5]
```diff
@@ -1,123 +0,0 @@
----
-# E2E test scenario using local dev images and manifests built from the source tree for following providers:
```
Comment (fabriziopandini) [inline test/e2e/config/docker-dev.yaml:2]: I'm dropping dev config, so we should not anymore put efforts in keeping it up-to-date

File: test/e2e/data/infrastructure-docker/metadata.yaml [hunk 6]
```diff
@@ -1,9 +0,0 @@
-apiVersion: clusterctl.cluster.x-k8s.io/v1alpha3
```
Comment (fabriziopandini) [inline test/e2e/data/infrastructure-docker/metadata.yaml:1]: Dropping this file because now it is embedded in clusterctl

File: test/e2e/kcp_adoption.go [hunk 7]
```diff
@@ -131,21 +132,24 @@ func KCPAdoptionSpec(ctx context.Context, inputGetter func() KCPAdoptionSpecInpu
...
-		controlPlane := framework.GetKubeadmControlPlaneByCluster(ctx, framework.GetKubeadmControlPlaneByClusterInput{
-			Lister:      client,
-			ClusterName: clusterName,
-			Namespace:   namespace.Name,
-		})
-		Expect(controlPlane).ToNot(BeNil())
+		Expect(input.BootstrapClusterProxy.ApplyWithArgs(ctx, workloadClusterTemplate, "--selector", "kcp-adoption.step2")).ShouldNot(HaveOccurred())
+
+		var controlPlane *controlplanev1.KubeadmControlPlane
+		Eventually(func() *controlplanev1.KubeadmControlPlane {
```
Comment (fabriziopandini) [inline test/e2e/kcp_adoption.go:138]: Adding Eventually to fix some flakes observed during local tests

**Comment by fabriziopandini** [comment]
/hold 
investigating an error on ci-e2e.sh that shows up un CI only, while locally everything works just fine 🤔

File: scripts/ci-e2e.sh [hunk 8]
```diff
@@ -31,7 +31,7 @@ source "${REPO_ROOT}/hack/ensure-kustomize.sh"
 # Configure provider images generation;
 # please ensure the generated image name matches image names used in the E2E_CONF_FILE
 export REGISTRY=gcr.io/k8s-staging-cluster-api
-export TAG=ci
```
Comment (vincepri) [inline scripts/ci-e2e.sh:34]: Why change the tag for the CI?

File: test/e2e/Makefile [hunk 9]
```diff
@@ -28,7 +28,7 @@ export GOPROXY
 REPO_ROOT := $(shell git rev-parse --show-toplevel)
 
 help:  ## Display this help
-	@awk 'BEGIN {FS = ":.*##"; printf "\nUsage:\n  make \033[36m<target>\033[0m\n"} /^[a-zA-Z_-]+:.*?##/ { printf "  \033[36m%-15s\033[0m %s\n", $$1, $$2 } /^##@/ { printf "\n\033[1m%s\033[0m\n", substr($$0, 5) } ' $(MAKEFILE_LIST)
```
Comment (vincepri) [inline test/e2e/Makefile:31]: Was this change intentional? What does it do?

**Comment by vincepri** [comment]
From the PR's description, seems we're now splitting manifests in order to separate MHC components? Would this be something that we would expect users to do as well? 

During move, we should be able to pause controllers and look at the state of the cluster if it's safe to move, is there anything stopping us to do so?

Can we split this PR in multiple ones? In particular:
> also, in order to simplify things, this PR drops the distinction between CI and DEV test config
> On top of that, this PR includes, a set of nits for improving test logs (mostly uppercase)

These changes seems outside of the PR's scope

File: test/e2e/config/docker.yaml [hunk 10]
```diff
@@ -5,17 +5,17 @@
 # - control-plane kubeadm
 # - docker
 
-# For creating local dev images run ./scripts/ci-e2e.sh
+# For creating local dev images run make docker-build-e2e from the main CAPI repository
```
Comment (sedefsavas) [inline test/e2e/config/docker.yaml:8]: I think we can keep `run ./scripts/ci-e2e.sh` here, because running `make docker-build-e2e` will not create images with the tag `dev` this file is using.

File: test/e2e/data/infrastructure-docker/cluster-template/kustomization.yaml [hunk 11]
```diff
@@ -0,0 +1,5 @@
+bases:
```
Comment (sedefsavas) [inline test/e2e/data/infrastructure-docker/cluster-template/kustomization.yaml:1]: nit. We can generate cluster-template.yaml on the fly and use that as a base to avoid extra changes we will need to do here when apis change.

File: test/e2e/Makefile [hunk 12]
```diff
@@ -38,28 +38,43 @@ TOOLS_DIR := $(REPO_ROOT)/hack/tools
...
+$(KUSTOMIZE): # Build kustomize from tools folder.
+    cd $(TOOLS_DIR); go build -tags=tools -o $(BIN_DIR)/kustomize sigs.k8s.io/kustomize/kustomize/v3
+
 ## --------------------------------------
-## Testing
+## Templates
 ## --------------------------------------
 
-TEST_E2E_DIR := $(REPO_ROOT)/test/e2e
+DOCKER_TEMPLATES := $(REPO_ROOT)/test/e2e/data/infrastructure-docker
```
Comment (sedefsavas) [inline test/e2e/Makefile:53]: Can we put them under `templates` dir like other providers? kcp-adoption does not make much sense to put it under that though.

If we will keep them under `e2e/data` then we can modify `cluster-templates` below to `e2e-cluster-templates` to be explicit about their purpose.

File: test/framework/machine_helpers.go [hunk 13]
```diff
@@ -120,8 +121,7 @@ func WaitForControlPlaneMachinesToBeUpgraded(ctx context.Context, input WaitForC
 	Expect(input.KubernetesUpgradeVersion).ToNot(BeEmpty(), "Invalid argument. input.KubernetesUpgradeVersion can't be empty when calling WaitForControlPlaneMachinesToBeUpgraded")
 	Expect(input.MachineCount).To(BeNumerically(">", 0), "Invalid argument. input.MachineCount can't be smaller than 1 when calling WaitForControlPlaneMachinesToBeUpgraded")
 
-	By("ensuring all machines have upgraded kubernetes version")
-	log.Logf("Ensuring all MachineDeployment Machines have upgraded kubernetes version %s", input.KubernetesUpgradeVersion)
+	By(fmt.Sprintf("Ensuring all control-plane have upgraded kubernetes version %s", input.KubernetesUpgradeVersion))
```
Comment (sedefsavas) [inline test/framework/machine_helpers.go:124]:
```suggestion
	By(fmt.Sprintf("Ensuring all control-plane machines have upgraded kubernetes version %s", input.KubernetesUpgradeVersion))
```

**Review by sedefsavas — COMMENTED** [review]
Thanks for the fix!

This PR is also addressing this issue: https://github.com/kubernetes-sigs/cluster-api/issues/3461
I am fine with adding it in the title and the issues fixed by this PR or making this separate PRs.

**Comment by sedefsavas** [comment]
@fabriziopandini Locally, works for me as well.


/test pull-cluster-api-e2e

**Comment by sedefsavas** [comment]
In `test/e2e/Makefile`

    cd $(TOOLS_DIR); go build -tags=tools -o $(BIN_DIR)/kustomize sigs.k8s.io/kustomize/kustomize/v3

should be 

	cd $(TOOLS_DIR) && go build -tags=tools -o $(BIN_DIR)/kustomize sigs.k8s.io/kustomize/kustomize/v3

File: test/framework/machine_helpers.go (hunk 13 above)
Comment (fabriziopandini) [reply test/framework/machine_helpers.go:124]: fixed in https://github.com/kubernetes-sigs/cluster-api/pull/3649

File: scripts/ci-e2e.sh (hunk 8 above)
Comment (fabriziopandini) [reply scripts/ci-e2e.sh:34]: moved to https://github.com/kubernetes-sigs/cluster-api/pull/3651

File: test/e2e/Makefile (hunk 9 above)
Comment (fabriziopandini) [reply test/e2e/Makefile:31]: moved to #3651

File: test/e2e/config/docker.yaml (hunk 10 above)
Comment (fabriziopandini) [reply test/e2e/config/docker.yaml:8]: moved to #3651

File: test/e2e/data/infrastructure-docker/cluster-template/kustomization.yaml (hunk 11 above)
Comment (fabriziopandini) [reply test/e2e/data/infrastructure-docker/cluster-template/kustomization.yaml:1]: moved to #3651

File: test/e2e/Makefile (hunk 12 above)
Comment (fabriziopandini) [reply test/e2e/Makefile:53]: moved to #3651

**Comment by fabriziopandini** [comment]
@vincepri 

> From the PR's description, seems we're now splitting manifests in order to separate MHC components? Would this be something that we would expect users to do as well?

No. The current MHC object in E2E tests is specifically configured to always trigger remediation after 30s the node is started, and I removed it from the other e2e tests in order to avoid flakes due to remediation kicking in case of slow execution

> During move, we should be able to pause controllers and look at the state of the cluster if it's safe to move, is there anything stopping us to do so?

This is not how the move logics works. Currently we are pausing reconciliation on clusters included in the Move scope, but the controller will continue to run.

> Can we split this PR in multiple ones? 

Done! rif 
https://github.com/kubernetes-sigs/cluster-api/pull/3649
https://github.com/kubernetes-sigs/cluster-api/pull/3650
https://github.com/kubernetes-sigs/cluster-api/pull/3651

> These changes seems outside of the PR's scope

Some of them are nits, but refactoring templates for MHC is potentially related to self-hosted flakes, see the note above about MHC and https://github.com/kubernetes-sigs/cluster-api/issues/3589#issuecomment-688333335.
However, no problem to move to another PR

**Comment by vincepri** [comment]
> No. The current MHC object in E2E tests is specifically configured to always trigger remediation after 30s the node is started, and I removed it from the other e2e tests in order to avoid flakes due to remediation kicking in case of slow execution

I assume this change has moved to the other PR, I'll take a look there before commenting

> This is not how the move logics works. Currently we are pausing reconciliation on clusters included in the Move scope, but the controller will continue to run.

We should probably consider stopping all controllers before move, for safety purposes and to avoid any running slow worker to perform actions while move is running.

**Comment by randomvariable** [comment]
/lgtm

pending e2e pass

**Review by vincepri — APPROVED** [review]
/approve
/lgtm
